src/filters/detectors/assurance.cpp

```diff
@@ -9,71 +10,174 @@
 namespace mesos {
 namespace serenity {
 
-Result<Detection> AssuranceDetector::processSample(
-    double_t in) {
-  this->window.push_back(in);
+void AssuranceDetector::shiftBasePoints() {
+  for (std::list<double_t>::iterator& basePoint : this->basePoints) {
+    basePoint++;
+  }
+}
+
 
-  if (this->window.size() < this->cfg.getU64(detector::WINDOW_SIZE)) {
-    return None();  // Only warm up.
+Detection AssuranceDetector::createContention(double_t severity = -1) {
+  Detection cpd;
+  if (severity > 0) {
+    cpd.severity = severity;
   }
 
-  double_t basePoint = this->window.front();
+  SERENITY_LOG(INFO) << " Created contention with severity = "
+    << (cpd.severity.isSome() ? std::to_string(cpd.severity.get()) : "<none>");
+  return cpd;
+}
+
+
+void AssuranceDetector::recalculateParams() {
+  this->window.clear();
+  this->basePoints.clear();
+
+  uint64_t windowSize = this->cfg.getU64(detector::WINDOW_SIZE);
+
+  // Find the biggest n in the T-2^n which fits within window length.
+  uint64_t checkpoints = 0;
+  while (windowSize > 0) {
+    windowSize = windowSize >> 1;
+    checkpoints++;
+  }
+
+  // Make sure it does not exceed MAX_CHECKPOINSs option.
+  if (checkpoints > this->cfg.getU64(detector::MAX_CHECKPOINTS)) {
+    SERENITY_LOG(WARNING) << "Too wide windowSize.";
+    checkpoints = this->cfg.getU64(detector::MAX_CHECKPOINTS);
+  }
+
+  // Get the Quorum number from QUORUM fraction parameter.
```

Comment:
Which 'quorum' are you referring to here? I hope we can capture this in the PR comment